lib/doctest.js

```diff
@@ -200,17 +203,39 @@ function processLine(
       value = stripLeading (1, ' ', stripLeading (1, '>', trimmedLine));
       accum.tests.push ($test = {});
       $test[accum.state = INPUT] = {value: value};
+      $test[OUTPUT] = {values: []};
       input ($test);
-    } else if (trimmedLine.charAt (0) === '.') {
+    } else if (accum.state === INPUT && trimmedLine.charAt (0) === '.') {
       value = stripLeading (1, ' ', stripLeading (Infinity, '.', trimmedLine));
       $test = accum.tests[accum.tests.length - 1];
-      $test[accum.state].value += '\n' + value;
-      (accum.state === INPUT ? appendToInput : appendToOutput) ($test);
+      $test[INPUT].value += '\n' + value;
+      appendToInput ($test);
+    } else if (accum.state === OUTPUT && trimmedLine.charAt (0) === '.') {
+      value = stripLeading (1, ' ', stripLeading (Infinity, '.', trimmedLine));
+      $test = accum.tests[accum.tests.length - 1];
+      $test[OUTPUT].values[$test[OUTPUT].values.length - 1].value += '\n'
+                                                                  + value;
+      appendToOutput ($test);
+    } else if (MATCH_LOG.test (trimmedLine)) {
+      value = stripLeading (1, ' ', trimmedLine.replace (MATCH_LOG, ''));
+      $test = accum.tests[accum.tests.length - 1];
+      $test[accum.state = OUTPUT].values.push ({
+        channel: MATCH_LOG.exec (trimmedLine)[1],
+        value: value
+      });
+      if ($test[OUTPUT].values.length === 1) {
+        output ($test);
+      }
     } else if (accum.state === INPUT) {
       value = trimmedLine;
       $test = accum.tests[accum.tests.length - 1];
-      $test[accum.state = OUTPUT] = {value: value};
-      output ($test);
+      $test[accum.state = OUTPUT].values.push ({
+        channel: null,
+        value: value
+      });
+      if ($test[OUTPUT].values.length === 1) {
+        output ($test);
+      }
```

Comment:
What is the significance of `$test[OUTPUT].values` containing *exactly* one element? Will the array ever contain more than one element?

Reply:
It contains additional elements if the user specifies expected log lines. Eg.

```js
> ( log ('stdout') (1)
. , setTimeout (log ('stdout'), 1, 2)
. , log ('stderr') (3) )
[stdout]: 1
[stderr]: 3
undefined
[stdout]: 2
```

Here there's 4 elements collected into the outputs array.

Reply:
What would the value of `$test[OUTPUT].values` be in the example above?

Reply:
```js
[
  {channel: 'stdout', value: 1},
  {channel: 'stderr', value: 3},
  {channel: null, value: undefined},
  {channel: 'stdout', value: 2}
]
```

Reply:
The line numbers are not necessarily correct. Consider this simple test case:

```console
$ printf "// %s\n" "> log ('log') ('hello'), 2 + 2" "[log]: 'hello'" "5" >/tmp/index.js

$ cat /tmp/index.js
// > log ('log') ('hello'), 2 + 2
// [log]: 'hello'
// 5

$ bin/doctest --log-function log -- /tmp/index.js
running doctests in /tmp/index.js...
.x
FAIL: expected 5 on line 2 (got 4)
```

The error message erroneously refers to line 2.

Reply:
That's something I noticed as well. I think it's because the `loc` is still assigned to the outer structure, and the error message probably uses `loc.start`.

Reply:
So for this feature to work properly, the structure would have to change from something like:

```js
{
  values: [{channel: 'a', value 1}],
  loc: {start: {line: 1, column: 1}, end: {line: 1, column: 2}}
}
```

To something like:

```js
[{
  channel: 'a',
  value 1,
  loc: {start: {line: 1, column: 1}, end: {line: 1, column: 2}}
}]
```

Reply:
> The line numbers are not necessarily correct.

There might be a misunderstanding here. The values `1`, `3`, and `2` were not referring to line numbers, but to the values that were passed to `log`. The line numbers are tracked as a whole, as shown above, from the first line containing output, until the last line.

Reply:
> The line numbers are tracked as a whole, as shown above, from the first line containing output, until the last line.

Right. I don't find this behaviour intuitive, though. I would like to track the lines individually, as you proposed above. :)